radixdlt-core/radixdlt/src/main/java/com/radixdlt/statecomputer/RadixEngineMempool.java

```diff
@@ -101,62 +100,55 @@ public void add(Command command) throws MempoolRejectedException {
 			radixEngine.deleteBranches();
 		}
 
-		this.data.put(atom.getAID(), command);
+		this.data.put(command, atom);
 
 		atom.uniqueInstructions()
-			.forEach(i -> particleIndex.merge(i, Set.of(atom.getAID()), Sets::union));
+			.forEach(i -> particleIndex.merge(i, Set.of(command), Sets::union));
 
 		updateCounts();
 	}
 
 	@Override
-	public List<Pair<Atom, Exception>> committed(List<Atom> commands) {
-		commands.forEach(a -> data.remove(a.getAID()));
-		final List<Pair<Atom, Exception>> removed = new ArrayList<>();
-		commands.stream()
-			.flatMap(Atom::uniqueInstructions)
+	public List<Pair<Command, Exception>> committed(List<Atom> atoms) {
+		final List<Pair<Command, Exception>> removed = new ArrayList<>();
+		atoms.forEach(atom -> atom.uniqueInstructions()
 			.flatMap(p -> {
-				Set<AID> aids = particleIndex.remove(p);
-				return aids != null ? aids.stream() : Stream.empty();
-			})
-			.forEach(aid -> {
-				var command = data.remove(aid);
+				Set<Command> cmds = particleIndex.remove(p);
+				return cmds != null ? cmds.stream() : Stream.empty();
+			}).forEach(cmd -> {
+				var toRemove = data.remove(cmd);
 				// TODO: Cleanup
-				if (command != null) {
-					Atom atom;
-					try {
-						atom = DefaultSerialization.getInstance().fromDson(command.getPayload(), Atom.class);
-					} catch (DeserializeException e) {
-						throw new IllegalStateException();
-					}
-
-					removed.add(Pair.of(atom, new RadixEngineMempoolException(
+				if (toRemove != null && !toRemove.equals(atom)) {
```

Comment:
Perhaps it worth to reverse the comparison: `atom.equals(toRemove)`. This should make comparison safe even if `data.remove(command)` will return `null`.